src/dotnet/CarbonAware.Tools.WattTimeClient/WattTimeClient.cs

```diff
@@ -0,0 +1,327 @@
+
+        var response = await this.client.GetAsync(uriPath);
+
+        if (!response.IsSuccessStatusCode && RetriableStatusCodes.Contains(response.StatusCode))
+        {
+            Log.LogDebug("Failed to get url {url} with status code {statusCode}.  Attempting to log in again.", uriPath, response.StatusCode);
+            await this.UpdateAuthTokenAsync();
+            response = await this.client.GetAsync(uriPath);
+        }
+
+        if (!response.IsSuccessStatusCode)
+        {
+            Log.LogError("Error getting data from WattTime.  StatusCode: {statusCode}. Response: {response}", response.StatusCode, response);
+
+            throw new WattTimeClientException($"Error getting data from WattTime: {response.StatusCode}", response);
+        }
+
+        return response;
+    }
+
+    private async Task<string> GetAsyncStringWithAuthRetry(string uriPath)
+    {
+        var response = await this.GetAsyncWithAuthRetry(uriPath);
+        var data = response.Content.ReadAsStringAsync();
+        return data.Result ?? string.Empty;
+    }
+
+    private async Task<Stream> GetAsyncStreamWithAuthRetry(string uriPath)
+    {
+        var response = await this.GetAsyncWithAuthRetry(uriPath);
+        return await response.Content.ReadAsStreamAsync();
+    }
+
+
+    private async Task EnsureTokenAsync()
+    {
+        if (this.client.DefaultRequestHeaders.Authorization == null)
+        {
+            await this.UpdateAuthTokenAsync();
+        }
+    }
+
+    private async Task UpdateAuthTokenAsync()
+    {
+        using (var activity = ActivitySource.StartActivity())
+        {
+            activity?.SetTag(QueryStrings.Username, this.Configuration.Username);
+
+            Log.LogInformation("Attempting to log in user {username}", this.Configuration.Username);
+
+            this.SetBasicAuthenticationHeader();
+            var response = await this.client.GetAsync(Paths.Login);
+
+            LoginResult? data = null;
+
+            if (response.IsSuccessStatusCode)
+            {
+                var json = await response.Content.ReadAsStringAsync() ?? String.Empty;
+
+                data = JsonSerializer.Deserialize<LoginResult>(json, options);
+            }
+
+            if (data == null)
+            {
+                Log.LogError("Login failed for user {username}.  Response: {response}", this.Configuration.Username, response);
+                throw new WattTimeClientException("Login failed.", response);
+            }
+
+            this.SetBearerAuthenticationHeader(data.Token);
+        }
+    }
+
+    private void SetBasicAuthenticationHeader()
+    {
+        var authToken = Encoding.ASCII.GetBytes($"{this.Configuration.Username}:{this.Configuration.Password}");
+        this.client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(AuthenticationHeaderTypes.Basic, Convert.ToBase64String(authToken));
+    }
+
+    internal void SetBearerAuthenticationHeader(string token)
```

Comment:
any preference to use `internal` here instead of `private`. Like `SetBasicAuth...`

Reply:
This is used for unit testing.